cmd/clusterctl/client/cluster/mover.go

```diff
@@ -595,6 +610,49 @@ func setClusterClassPause(proxy Proxy, clusterclasses []*node, pause bool, dryRu
 	return nil
 }
 
+func waitReadyForMove(proxy Proxy, nodes []*node, dryRun bool, backoff wait.Backoff) error {
+	if dryRun {
+		return nil
+	}
+
+	log := logf.Log
+
+	c, err := proxy.NewClient()
+	if err != nil {
+		return errors.Wrap(err, "error creating client")
+	}
+
+	for _, n := range nodes {
+		obj := &metav1.PartialObjectMetadata{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      n.identity.Name,
+				Namespace: n.identity.Namespace,
+			},
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: n.identity.APIVersion,
+				Kind:       n.identity.Kind,
+			},
+		}
+		key := client.ObjectKeyFromObject(obj)
+
+		if err := retryWithExponentialBackoff(backoff, func() error {
```

Comment:
As it is implemented the backoff applies to every CR, thus assuming that we are moving 100 objects, this can result in 100*5sec, which is a lot. Is that the expected behavior?

Reply:
That's what I had in mind for the time being. I generally wouldn't expect this to ever wait for `N resources * timeout for one resource` since while `clusterctl` is waiting for one resource to be paused, the provider will continue doing work out-of-band to pause that and other resources in parallel. Overall I think the total time this waits will trend toward the longest wait for an individual resource in that case.

That same parallelism could be applied to the waits here as well, but that didn't seem worth the additional complexity under the assumption that resources are going to be paused in parallel.

Reply:
I got your point, It is not ideal because ultimately the time the users have to wait can vary in a significant way, but I'm with this for the first iteration if we can get a better UX / like suggested in other comments.

TL;DR;  if there is a block on an object, before starting the wait loop, we should log "Waiting for the block to be removed on object..." or something similar that makes the users aware of where the process is blocked

Reply:
Based on my read of the code, it looks like the "resource is not ready to move" error returned here will get logged here which happens before any `time.Sleep`:
https://github.com/kubernetes-sigs/cluster-api/blob/8f173a150434d735409beed2cec58a345926e29a/cmd/clusterctl/client/cluster/client.go#L234

Does that cover what you describe?

Reply:
Not really, this is the debug of every retry and it is very noisy; it will also be printed for every object, not only the objects with an actual block move.

So it does not comply with 

> if (and only if) there is a block on an object, before starting the wait loop we should log "Waiting for the block to be removed on ..." or something similar that makes the users aware of where the move process is blocked at

It just provides debug/trace details on every step of the wait loop when you are already in

Reply:
So you'd like to see a single log message indicating that at least one object is blocking the move?

Reply:
obj1 no block move annotation --> no message at all (and possibly skip entirely the wait loop)
obj2 block move annotation --> before starting the wait loop, "Waiting for the block to be removed on ..."
obj3-50 no block move annotation --> no message at all (and possibly skip entirely the wait loop)
obj51 block move annotation --> before starting the wait loop, "Waiting for the block to be removed on ..."
and so on

Reply:
That makes sense, I've updated this to do that. FWIW we need to enter the wait loop no matter what to determine if the annotation exists, but the loop is exited as soon as the annotation doesn't exist (which may be in the first iteration). You already suggested gathering whether the resource is blocked during object discovery which could make this short-circuit earlier, but I'd like to do that as a follow-up if that's still acceptable.

Reply:
> I'd like to do that as a follow-up if that's still acceptable

this would also be good to track in an issue

Reply:
Opened #9196